### PR TITLE
GAP-2111: Reorder questions

### PR DESCRIPTION
**Pull request by dominicwest**
## Description

Ticket [GAP-2111](https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2111)

Summary of the changes and the related issue. List any dependencies that are required for this change:
- New endpoint for re-ordering questions within a section within an application form by a given increment
- Unsure how to unit test the client side js behaviour as this is new for me, if anyone has experience plz educate me
- https://github.com/cabinetoffice/gap-find-admin-backend/pull/171

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End-to-End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

https://github.com/cabinetoffice/gap-find-admin-backend/assets/101722961/5241076e-f046-444d-ad62-36eb0dcabd81

# Checklist:

- [x] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation where applicable.
- [ ] I have run cypress tests, and they all pass locally.

[GAP-2111]: https://technologyprogramme.atlassian.net/browse/GAP-2111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx [hunk 1]
```diff
@@ -140,50 +212,66 @@ const EditSectionPage = ({
...
+          <div className="govuk-!-text-align-right govuk-width-container govuk-!-padding-top-4 govuk-!-padding-right-2">
+            <div className="govuk-grid-row">
+              <button
+                className={`govuk-button govuk-button--secondary govuk-!-margin-right-2 govuk-!-margin-bottom-0 ${styles['button']}`}
+                data-module="govuk-button"
+                data-cy="cyUpButton"
+                name={`Up/${question.questionId}`}
+                disabled={questionIndex === 0}
+                onClick={handleOnUpDownButtonClick}
+                data-testid={`upButton-${questionIndex}`}
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:246]: Doesn't look like we're using these attributes?

Comment (dominicwest) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:246]: Which ones? Name, disabled, onClick & data-testid are all used

Comment (jgunnCO) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:246]: Oh wait I see they're for the RTL stuff. I'm not a huge fan of having two different test ID attributes, could you not as easily do `getAllByRole` and use the array indexes to assert on which buttons are enabled? Eg:

```js
const upButtons = screen.getAllByRole('button', { name: 'Move section up' });
expect(upButtons[0]).toBeDisabled();
expect(upButtons[1]).not.toBeDisabled();
expect(upButtons[2]).not.toBeDisabled();

```

File: packages/admin/src/services/ApplicationService.ts [hunk 2]
```diff
@@ -106,11 +106,26 @@ const handleSectionOrdering = async (
   );
 };
 
+const handleQuestionOrdering = async (
+  sessionId: string,
+  applicationId: string,
+  sectionId: string,
+  questionId: string,
+  increment: number
+) => {
```
Comment (jgunnCO) [inline packages/admin/src/services/ApplicationService.ts:115]: If you're passing this many arguments into a function it's generally easier to just have one arg that's an object with all these properties - makes it way easier to add and remove args, and prevents mistakes with ordering the arguments correctly.

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx (hunk 1 above)
Comment (dominicwest) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:246]: Yeah sure - will give it a go

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx [hunk 3]
```diff
@@ -45,7 +86,29 @@ const EditSectionPage = ({
...
+  );
+  const formAction =
+    resolvedUrl.split('?')[0] + `?scrollPosition=${newScrollPosition}`;
+
+  function handleOnUpDownButtonClick() {
+    setNewScrollPosition(window.scrollY);
+    formRef?.current?.submit();
+  }
+
+  useEffect(() => {
```
Comment (jgunnCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:105]: I think scrolling on load is an actual use case for `useLayoutEffect` - it'll ensure the scroll completes synchronously before the page is rendered, which will prevent any weird flickering behaviour where the page loads for a second and _then_ scrolls.

Comment (dominicwest) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:105]: Ohhhh neat will test this out

File: packages/admin/src/services/ApplicationService.ts (hunk 2 above)
Comment (dominicwest) [reply packages/admin/src/services/ApplicationService.ts:115]: Agreed

Comment (dominicwest) [reply packages/admin/src/services/ApplicationService.ts:115]: Realised I have no tests for the post behaviour so I'll add these too

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx [hunk 4]
```diff
@@ -1,18 +1,56 @@
...
+        sessionId,
+        applicationId,
+        sectionId,
+        questionId,
+        increment
+      );
+    },
+    `/build-application/${applicationId}/${sectionId}?scrollPosition=${scrollPosition}`,
+    generateErrorPageParams(
+      'Something went wrong while trying to update section orders.',
```
Comment (ConorFayleTCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:46]:
```suggestion
      'Something went wrong while trying to update section order.',
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx [hunk 5]
```diff
@@ -140,50 +212,66 @@ const EditSectionPage = ({
...
+        ),
+        className: 'govuk-!-padding-1',
+      },
+      {
+        content: <p>{ResponseTypeLabels[question.responseType]}</p>,
+        className: 'govuk-!-padding-1',
+      },
+      {
+        content: (
+          <div className="govuk-!-text-align-right govuk-width-container govuk-!-padding-top-4 govuk-!-padding-right-2">
```
Comment (ConorFayleTCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/index.page.tsx:237]: could you extract out this component? would make it easier to read, less nesting etc

File: packages/gap-web-ui/src/utils/UnitTestHelpers.tsx [hunk 6]
```diff
@@ -38,6 +38,9 @@ const getContext = (
         cookies: { 'gap-test': 'testSessionId' },
         csrfToken: () => 'testCSRFToken',
```
Comment (jgunnCO) [inline packages/gap-web-ui/src/utils/UnitTestHelpers.tsx:39]: can get rid of this line now as it's just a red herring at this point

**Review by john-tco — APPROVED** [review]

**Pull request merged**